psamm/datasource/native.py

```diff
@@ -174,28 +174,44 @@ class NativeModel(object):
     directory subtree that specifies part of the model.
     """
 
-    def __init__(self, path):
-        """Create a model from the specified model file or directory"""
+    def __init__(self, model_from, context=None):
+        """Create a model from the specified model file.
 
+        Model can be a directory, long string, open file, or dictionary.
+        """
+        if isinstance(model_from, string_types):
+            self._model = yaml_load(model_from)
+            self._context = context
+        elif isinstance(model_from, dict):
+            self._context = context
+            self._model = model_from
+        elif hasattr(model_from, 'read') and callable(model_from.read):
+            self._context = context
+            self._model = yaml_load(model_from)
+
+    @classmethod
+    def load_model_from_path(cls, path):
         if os.path.isfile(path):
-            self._context = FilePathContext(path)
-            with open(self._context.filepath, 'r') as f:
-                self._model = yaml_load(f)
-        else:
+            context = FilePathContext(path)
+            with open(context.filepath, 'r') as f:
+                model = NativeModel(f, context)
+        elif os.path.isdir(path):
             # Try to open the default file
             for filename in DEFAULT_MODEL:
                 try:
-                    self._context = FilePathContext(
+                    context = FilePathContext(
                         os.path.join(path, filename))
-                    with open(self._context.filepath, 'r') as f:
-                        self._model = yaml_load(f)
+                    with open(context.filepath, 'r') as f:
+                        model = NativeModel(f, context)
                         break
                 except:
-                    logger.debug('Failed to load model file', exc_info=True)
+                    logger.debug('Failed to load model file',
+                                 exc_info=True)
             else:
                 # No model could be loaded
-                raise ParseError('No model file could be found ({})'.format(
+                raise ParseError('No model file could be found({})'.format(
                     ', '.join(DEFAULT_MODEL)))
+        return model
```

Comment:
Same here: what happens if the provided `path` does not exist? It will probably return `None` but it should preferably raise an exception.